### PR TITLE
Add feature to upload large submisison and annotation files.

### PR DESCRIPTION
**Pull request by KhalidRmb**
Add support for uploading large files.

### CODE REVIEW

File: evalai/utils/submissions.py [hunk 1]
```diff
@@ -18,6 +18,86 @@
...
+        )
+        presigned_url = response.data.get("presigned_url")
+        submission_message = response.data.get("submission_message")
+
+        response = requests.post(
+            presigned_url, 
+            files=file,
+        )
+
+        if response.status_code == 200:
```
Comment (Ram81) [inline evalai/utils/submissions.py:45]: can we use HttpStatus.ok instead of hardcoding 200

File: evalai/utils/submissions.py [hunk 2]
```diff
@@ -18,6 +18,86 @@
...
 
+def upload_submission_file_with_presigned_url(challenge_pk, challenge_phase_pk, data, file):
+    """
+    Function to upload a file to AWS using a presigned url
+    """
+    url = "{}{}".format(get_host_url(), URLS.get_presigned_url_for_annotations.value)
+    url = url.format(challenge_pk, challenge_phase_pk)
+
+    headers = get_request_header()
+    input_file = {"input_file": }  # File should be a dummy file here.
```
Comment (Ram81) [inline evalai/utils/submissions.py:29]: can you add a default value

**Comment by KhalidRmb** [comment]
@RishabhJain2018 @deshraj @Ram81 PTAL.

File: evalai/challenges.py [hunk 3]
```diff
@@ -234,4 +235,44 @@ def submit(ctx, file):
     make_submission(ctx.challenge_id, ctx.phase_id, file, submission_metadata)
 
 
+@phase.command(context_settings={"ignore_unknown_options": True})
+@click.pass_obj
+@click.option("--file", type=click.STRING, required=True, help="File path to the submission file")
+def upload_submission(ctx, file):
+    """
+    For uploading large submission files to S3.
```
Comment (Ram81) [inline evalai/challenges.py:243]: @KhalidRmb can you add proper docstrings.

File: evalai/challenges.py [hunk 4]
```diff
@@ -234,4 +235,44 @@ def submit(ctx, file):
...
+    """
+    """
+    Invoked by runing 'evalai challenge CHALLENGE phase PHASE upload_submission FILE'
```
Comment (Ram81) [inline evalai/challenges.py:246]: why are there two separate docstrings?

File: evalai/challenges.py [hunk 5]
```diff
@@ -234,4 +235,44 @@ def submit(ctx, file):
...
+        )
+    upload_submission_file_with_presigned_url(ctx.phase_id, file, submission_metadata)
+
+
+@phase.command(context_settings={"ignore_unknown_options": True})
+@click.pass_obj
+@click.option("--file", type=click.STRING, required=True, help="File path to the annotations file")
+def upload_annotation(ctx, file):
+    """
+    For uploading large test annotations for a challenge phase to S3.
```
Comment (Ram81) [inline evalai/challenges.py:270]: same here

File: evalai/utils/challenges.py [hunk 6]
```diff
@@ -667,3 +669,72 @@ def display_leaderboard(challenge_id, phase_split_id):
...
+        sys.exit(1)
+    echo(
+        style(
+            "\nYour annotation file {} for challenge phase {} is successfully uploaded.\n".format(
+                file, challenge_phase_pk
+            ),
+            fg="green",
+            bold=True,
+        )
+    )
```
Comment (Ram81) [inline evalai/utils/challenges.py:740]: add new line

File: evalai/utils/submissions.py [hunk 7]
```diff
@@ -18,6 +21,91 @@
 requests.packages.urllib3.disable_warnings()
 
 
+def upload_submission_file_with_presigned_url(challenge_phase_pk, file, submission_metadata={}):
+    """
+    Function to upload a file to AWS using a presigned url
```
Comment (Ram81) [inline evalai/utils/submissions.py:26]: same here for docstring

File: evalai/utils/submissions.py [hunk 8]
```diff
@@ -18,6 +21,91 @@
...
+            response.raise_for_status()
+
+        response = response.json()
+        presigned_url = response.get("presigned_url")
+        submission_pk = response.get("submission_pk")
+
+        # Uploading the submisison file to S3.
+        echo(
+            style(
+                "Your file is being uploaded...",
```
Comment (Ram81) [inline evalai/utils/submissions.py:49]: change this to `Uploading the submission file......` ?

File: evalai/utils/submissions.py [hunk 9]
```diff
@@ -18,6 +21,91 @@
...
+                    presigned_url,
+                    data=f
+                )
+                response.raise_for_status()
+            except requests.exceptions.HTTPError as err:
+                if response.status_code is not HTTPStatus.OK:
+                    echo("There was an error while uploading the file: {}".format(err))
+                    sys.exit(1)
+
+        # Publishing the submisison message, for processing by the submission worker.
```
Comment (Ram81) [inline evalai/utils/submissions.py:66]: change this to `Publishing submission message to the message queue for processing` ?

File: evalai/utils/submissions.py [hunk 10]
```diff
@@ -18,6 +21,91 @@
...
+                " Please check the Host URL.\n",
+                bold=True,
+                fg="red",
+            )
+        )
+        sys.exit(1)
+    response = response.json()
+    echo(
+        style(
+            "\nYour file {} with the ID {} is successfully submitted for evaluation.\n".format(
```
Comment (Ram81) [inline evalai/utils/submissions.py:100]: Change this to `Your submission {} with the submission id {} is successfully submitted for evaluation.` ?

File: evalai/utils/challenges.py [hunk 11]
```diff
@@ -667,3 +669,72 @@ def display_leaderboard(challenge_id, phase_split_id):
...
+        response = requests.get(url, headers=headers, data=data)
+        if response.status_code is not HTTPStatus.OK:
+            response.raise_for_status()
+
+        response = response.json()
+
+        # Uploading the annotation file for the current phase to S3.
+        echo(
+            style(
+                "Your file is being uploaded...",
```
Comment (Ram81) [inline evalai/utils/challenges.py:691]: change this to `Uploading the test annotations.....`? or you can change it to `Uploading the {phase_name} phase annotations...`

File: evalai/utils/common.py [hunk 12]
```diff
@@ -31,6 +34,36 @@ def convert(self, value, param, ctx):
             )
 
 
+def upload_with_presigned_url(file, presigned_url):
+    """Function to upload a file, given the target presigned s3 url.
```
Comment (Ram81) [inline evalai/utils/common.py:38]: Fix the docstring

File: evalai/utils/common.py [hunk 13]
```diff
@@ -31,6 +34,36 @@ def convert(self, value, param, ctx):
             )
 
 
+def upload_with_presigned_url(file, presigned_url):
```
Comment (Ram81) [inline evalai/utils/common.py:37]: if `file` is the file_path please rename variables accordingly

File: evalai/utils/common.py [hunk 14]
```diff
@@ -31,6 +34,36 @@ def convert(self, value, param, ctx):
...
+
+    with open(os.path.realpath(file), 'rb') as f:
+        try:
+            response = requests.put(
+                presigned_url,
+                data=f
+            )
+            return response
+        except Exception as err:
+            echo("There was some error while uploading the file: {}".format(err))
```
Comment (Ram81) [inline evalai/utils/common.py:60]: change to `There was an error while uploading the submission file:`

File: evalai/utils/common.py [hunk 15]
```diff
@@ -31,6 +34,36 @@ def convert(self, value, param, ctx):
...
+            sys.exit(1)
+        if response.status_code is not HTTPStatus.OK:
+            echo("There was some error while uploading the file: ")
```
Comment (Ram81) [inline evalai/utils/common.py:63]: same here

File: evalai/utils/submissions.py [hunk 16]
```diff
@@ -4,20 +4,90 @@
...
 )
 
 
 requests.packages.urllib3.disable_warnings()
 
 
+def upload_submission_file_with_presigned_url(challenge_phase_pk, file, submission_metadata={}):
+    """
+    Function to upload a file to AWS using a presigned url
+    """
```
Comment (Ram81) [inline evalai/utils/submissions.py:26]: @KhalidRmb docstring is not fixed here.

File: evalai/utils/submissions.py [hunk 17]
```diff
@@ -4,20 +4,90 @@
...
+            url, headers=headers, data=data
+        )
+        if response.status_code is not HTTPStatus.CREATED:
+            response.raise_for_status()
+
+        response = response.json()
+        presigned_url = response.get("presigned_url")
+        submission_pk = response.get("submission_pk")
+
+        # Uploading the submisison file to S3.
```
Comment (Ram81) [inline evalai/utils/submissions.py:45]: be consisten either use `.` for every comment or don't use for anyone

File: evalai/utils/challenges.py [hunk 18]
```diff
@@ -667,3 +669,54 @@ def display_leaderboard(challenge_id, phase_split_id):
...
+                "\nCould not establish a connection to EvalAI."
+                " Please check the Host URL.\n",
+                bold=True,
+                fg="red",
+            )
+        )
+        sys.exit(1)
+    echo(
+        style(
+            "\nYour annotation file {} for challenge phase {} is successfully uploaded.\n".format(
```
Comment (Ram81) [inline evalai/utils/challenges.py:716]: do we need `Your` shouldn't this be `The annotation file` and so on?

File: evalai/challenges.py [hunk 19]
```diff
@@ -234,4 +235,46 @@ def submit(ctx, file):
...
+            style("Publication URL", fg="yellow"), type=str, default=""
+        )
+    upload_submission_file_with_presigned_url(ctx.phase_id, file, submission_metadata)
+
+
+@phase.command(context_settings={"ignore_unknown_options": True})
+@click.pass_obj
+@click.option("--file", type=click.STRING, required=True, help="File path to the annotations file")
+def upload_annotation(ctx, file):
+    """For uploading large test annotations for a challenge phase to S3.
```
Comment (Ram81) [inline evalai/challenges.py:270]: fix docstring

File: evalai/challenges.py [hunk 20]
```diff
@@ -234,4 +235,46 @@ def submit(ctx, file):
     make_submission(ctx.challenge_id, ctx.phase_id, file, submission_metadata)
 
 
+@phase.command(context_settings={"ignore_unknown_options": True})
+@click.pass_obj
+@click.option("--file", type=click.STRING, required=True, help="File path to the submission file")
+def upload_submission(ctx, file):
+    """For uploading large submission files to S3.
```
Comment (Ram81) [inline evalai/challenges.py:242]: fix docstring

File: evalai/utils/challenges.py [hunk 21]
```diff
@@ -667,3 +669,54 @@ def display_leaderboard(challenge_id, phase_split_id):
...
+        presigned_url = response.get("presigned_url")
+
+        # Uploading the annotation file for the current phase to S3.
+        response = upload_with_presigned_url(file, presigned_url)
+    except requests.exceptions.HTTPError as err:
+        if response.status_code in EVALAI_ERROR_CODES:
+            validate_token(response.json())
+            echo(
+                style(
+                    "Error: {}".format(response.json()["error"]),
```
Comment (Ram81) [inline evalai/utils/challenges.py:696]: shouldn't this message be more descriptive? similar to what was used for submission error message

File: evalai/utils/common.py (hunk 12 above)
Comment (KhalidRmb) [reply evalai/utils/common.py:38]: Fix how? Move it to the next line? I'm following this, as Rishabh suggested this in EvalAI repository for my other PR: https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard

Comment (Ram81) [reply evalai/utils/common.py:38]: See this for an example https://github.com/Cloud-CV/EvalAI/blob/master/apps/base/utils.py#L152

File: evalai/challenges.py [hunk 22]
```diff
@@ -234,4 +235,48 @@ def submit(ctx, file):
     make_submission(ctx.challenge_id, ctx.phase_id, file, submission_metadata)
 
 
+@phase.command(context_settings={"ignore_unknown_options": True})
+@click.pass_obj
+@click.option("--file", type=click.STRING, required=True, help="File path to the submission file")
+def upload_submission(ctx, file_name):
+    """
+    For uploading large submission files to S3
+    Invoked by runing 'evalai challenge CHALLENGE phase PHASE upload_submission FILE'
```
Comment (Ram81) [inline evalai/challenges.py:244]: typo in `running`

File: evalai/challenges.py [hunk 23]
```diff
@@ -234,4 +235,48 @@ def submit(ctx, file):
...
+    upload_submission_file_with_presigned_url(ctx.phase_id, file_name, submission_metadata)
+
+
+@phase.command(context_settings={"ignore_unknown_options": True})
+@click.pass_obj
+@click.option("--file", type=click.STRING, required=True, help="File path to the annotations file")
+def upload_annotation(ctx, file_name):
+    """
+    For uploading large test annotations for a challenge phase to S3
+    Invoked by runing 'evalai challenge CHALLENGE phase PHASE upload_annotations FILE'
```
Comment (Ram81) [inline evalai/challenges.py:273]: same here

File: evalai/challenges.py [hunk 24]
```diff
@@ -234,4 +235,48 @@ def submit(ctx, file):
...
+
+    Arguments:
+        ctx (class click.Context) --  The context object which holds state of the invocation
+        file_name (str) -- the path of the file to be uploaded
```
Comment (Ram81) [inline evalai/challenges.py:277]: is this file name or file path? if param is path can we rename it to `file_path` if not can you change the description?

File: evalai/utils/challenges.py [hunk 25]
```diff
@@ -667,3 +669,62 @@ def display_leaderboard(challenge_id, phase_split_id):
...
         echo(style("Sorry, no Leaderboard results found.", bold=True, fg="red"))
+
+
+def upload_annotations_file_with_presigned_url(challenge_phase_pk, file_name):
+    """
+    Function to upload a large test annotation file for a challenge phase through presigned urls
+
+    Arguments:
+        challenge_phase_pk (int) -- id of the challenge phase
+        file_name (str) -- the path of the file to be uploaded
```
Comment (Ram81) [inline evalai/utils/challenges.py:680]: same here

File: evalai/utils/submissions.py [hunk 26]
```diff
@@ -4,20 +4,95 @@
...
     validate_token,
     validate_date_format,
-    convert_UTC_date_to_local,
 )
 
 
 requests.packages.urllib3.disable_warnings()
 
 
+def upload_submission_file_with_presigned_url(challenge_phase_pk, file_name, submission_metadata={}):
```
Comment (Ram81) [inline evalai/utils/submissions.py:23]: change this to `upload_presigned_url_submission_file` ?

File: evalai/utils/common.py [hunk 27]
```diff
@@ -31,6 +34,37 @@ def convert(self, value, param, ctx):
             )
 
 
+def upload_with_presigned_url(file_name, presigned_url):
```
Comment (Ram81) [inline evalai/utils/common.py:37]: change this to `upload_presigned_url_file`?

File: evalai/utils/challenges.py [hunk 28]
```diff
@@ -667,3 +669,62 @@ def display_leaderboard(challenge_id, phase_split_id):
         pretty_print_leaderboard_data(attributes, results)
     else:
         echo(style("Sorry, no Leaderboard results found.", bold=True, fg="red"))
+
+
+def upload_annotations_file_with_presigned_url(challenge_phase_pk, file_name):
```
Comment (Ram81) [inline evalai/utils/challenges.py:674]: change this to `upload_presigned_url_annontations` or `upload_presigned_url_challenge_phase_annontations`?

**Comment by RishabhJain2018** [comment]
@KhalidRmb Is this ready to be merged?

File: evalai/challenges.py (hunk 24 above)
Comment (KhalidRmb) [reply evalai/challenges.py:277]: Changing this to `--file` for now.
This is simple and the meaning seems straightforward to the host. I have however changed the method signature to ensure code clarity (i.e, file_name -> file_path)

**Comment by KhalidRmb** [comment]
> @KhalidRmb Is this ready to be merged?
> cc: @Ram81
@RishabhJain2018 Yes

**Comment by Ram81** [comment]
@RishabhJain2018 I am testing this on my local with staging endpoints. I'll confirm once verified

File: evalai/utils/common.py [hunk 29]
```diff
@@ -31,6 +33,34 @@ def convert(self, value, param, ctx):
...
+    """
+    Function to upload a file, given the target presigned s3 url
+
+    Arguments:
+        file_name (str) -- the path of the file to be uploaded
+        presigned_url (str) -- the presigned url to upload the file on s3
+    """
+    echo(
+        style(
+            "Uploading the submission file...",
```
Comment (Ram81) [inline evalai/utils/common.py:46]: @KhalidRmb we are showing `Uploading the submission file...` message for both test annotations and submissions. Please fix this

Comment (KhalidRmb) [reply evalai/utils/common.py]: Fixed this @Ram81

File: evalai/utils/common.py [hunk 30]
```diff
@@ -31,6 +33,34 @@ def convert(self, value, param, ctx):
...
+    """
+    echo(
+        style(
+            "Uploading the file...",
+            fg="green",
+            bold=False,
+        )
+    )
+
+    with open(os.path.realpath(file_name), 'rb') as f:
```
Comment (Ram81) [inline evalai/utils/common.py:52]: @KhalidRmb can we rename `f` to `file`

File: evalai/utils/submissions.py [hunk 31]
```diff
@@ -4,20 +4,98 @@
...
+        echo(
+            style(
+                "\nCould not establish a connection to EvalAI."
+                " Please check the Host URL.\n",
+                bold=True,
+                fg="red",
+            )
+        )
+        sys.exit(1)
+    response = response.json()
```
Comment (Ram81) [inline evalai/utils/submissions.py:87]: is this being used anywhere? if not please remove it

File: evalai/utils/submissions.py [hunk 32]
```diff
@@ -4,20 +4,98 @@
...
+        # Publishing submission message to the message queue for processing
+        url = "{}{}".format(get_host_url(), URLS.send_submission_message.value)
+        url = url.format(challenge_phase_pk, submission_pk)
+        response = requests.post(
+            url,
+            headers=headers,
+        )
+        response.raise_for_status()
+    except requests.exceptions.HTTPError as err:
+        if response.status_code in EVALAI_ERROR_CODES:
```
Comment (Ram81) [inline evalai/utils/submissions.py:65]: have you tested this exception handling block explicitly? if you get an exception at line 41 will the response object be accessible here?I don't think it is global

**Review by Ram81 — CHANGES_REQUESTED** [review]

File: evalai/challenges.py [hunk 33]
```diff
@@ -234,4 +235,52 @@ def submit(ctx, file):
     make_submission(ctx.challenge_id, ctx.phase_id, file, submission_metadata)
 
 
+@phase.command(context_settings={"ignore_unknown_options": True})
+@click.pass_obj
+@click.option("--file", type=click.STRING, required=True, help="File path to the submission file")
+def upload_submission(ctx, file):
```
Comment (RishabhJain2018) [inline evalai/challenges.py:241]: @KhalidRmb Isn't it possible to use the previous command only for creating submissions. Having two different commands for the same action seems a bit weird to me.

You can add a parameter `-l` or `--large` and decide on the basis of that.
cc: @Ram81 @deshraj

**Review by RishabhJain2018 — CHANGES_REQUESTED** [review]

File: evalai/challenges.py (hunk 33 above)
Comment (KhalidRmb) [reply evalai/challenges.py:241]: Makes sense..Will add the change.

Comment (KhalidRmb) [reply evalai/challenges.py]: Done @RishabhJain2018

File: evalai/utils/urls.py [hunk 34]
```diff
@@ -30,3 +30,6 @@ class URLS(Enum):
     )
     download_file = "/api/jobs/submission_files/?bucket={}&key={}"
     phase_details_using_slug = "/api/challenges/phase/{}/"
+    get_annotation_file_presigned_url = "/api/challenges/phases/{}/get_annotation_file_presigned_url/"
```
Comment (RishabhJain2018) [inline evalai/utils/urls.py:33]: `get_presigned_url_for_annotation_file`

File: evalai/utils/urls.py [hunk 35]
```diff
@@ -30,3 +30,6 @@ class URLS(Enum):
...
+    get_submission_file_presigned_url = "/api/jobs/phases/{}/get_submission_file_presigned_url/"
```
Comment (RishabhJain2018) [inline evalai/utils/urls.py:34]: `get_presigned_url_for_submission_file`

**Comment by KhalidRmb** [comment]
@RishabhJain2018 Made the changes

File: evalai/challenges.py [hunk 36]
```diff
@@ -207,31 +207,50 @@ def participate(ctx, team):
...
+                style("Method Description", fg="yellow"), type=str, default=""
+            )
+            submission_metadata["project_url"] = click.prompt(
+                style("Project URL", fg="yellow"), type=str, default=""
+            )
+            submission_metadata["publication_url"] = click.prompt(
+                style("Publication URL", fg="yellow"), type=str, default=""
+            )
+        if large:
+            print("LJKMKNJONJUNJUNJUNIJNIBJU")
```
Comment (RishabhJain2018) [inline evalai/challenges.py:250]: Please remove this print statement.

**Review by RishabhJain2018 — CHANGES_REQUESTED** [review]
A minor change but the rest looks pretty good and clean.

**Comment by RishabhJain2018** [comment]
@KhalidRmb Can you please add the relevant test cases and fix the travis build.

**Comment by KhalidRmb** [comment]
@RishabhJain2018 Fixed the build

**Review by RishabhJain2018 — APPROVED** [review]
LGTM and quite clean.
@Ram81 Can you please test this once before I merge this?

**Comment by Ram81** [comment]
@KhalidRmb can you share the usage of new commands?

**Review by Ram81 — APPROVED** [review]

**Pull request merged**